ipn/ipnlocal/c2n.go

```diff
@@ -111,37 +120,36 @@ func (b *LocalBackend) handleC2N(w http.ResponseWriter, r *http.Request) {
 	}
 }
 
-func (b *LocalBackend) handleC2NUpdate(w http.ResponseWriter, r *http.Request) {
-	// TODO(bradfitz): add some sort of semaphore that prevents two concurrent
-	// updates, or if one happened in the past 5 minutes, or something.
+func (b *LocalBackend) handleC2NUpdateGet(w http.ResponseWriter, r *http.Request) {
+	b.logf("c2n: GET /update received")
 
-	// GET returns the current status, and POST actually begins an update.
-	if r.Method != "GET" && r.Method != "POST" {
-		http.Error(w, "bad method", http.StatusMethodNotAllowed)
-		return
-	}
+	res := b.newC2NUpdateResponse()
+	res.Started = b.c2nUpdateStarted()
 
-	// If NewUpdater does not return an error, we can update the installation.
-	// Exception: When version.IsMacSysExt returns true, we don't support that
-	// yet. TODO(cpalmer, #6995): Implement it.
-	//
-	// Note that we create the Updater solely to check for errors; we do not
-	// invoke it here. For this purpose, it is ok to pass it a zero Arguments.
-	prefs := b.Prefs().AutoUpdate()
-	_, err := clientupdate.NewUpdater(clientupdate.Arguments{})
-	res := tailcfg.C2NUpdateResponse{
-		Enabled:   envknob.AllowsRemoteUpdate() || prefs.Apply,
-		Supported: err == nil && !version.IsMacSysExt(),
+	resb, err := json.Marshal(res)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
 	}
+	b.logf("c2n: GET /update response: %s", resb)
+	w.Header().Set("Content-Type", "application/json")
+	w.Write(resb)
+}
 
+func (b *LocalBackend) handleC2NUpdatePost(w http.ResponseWriter, r *http.Request) {
+	b.logf("c2n: POST /update received")
+	res := b.newC2NUpdateResponse()
 	defer func() {
+		resb, err := json.Marshal(res)
```

Comment:
I think this (and the above) is probably a bit excessive; we control the contents of this type fully within the package. I'd probably simplify it to

```go
json.NewEncoder(w).Encode(res)
```

YMMV.

Reply:
I changed it be able to log the response body.
It's very handy to see what was returned in the local tailscaled logs :)

Reply:
That means we'll be shipping the full response to logtail, though—do we really want that in production? If you think so I don't object, but our logs are already pretty full of spam that was only useful during development. 🙂

Reply:
Hmm, good point, I didn't consider log shipping.
Even though the response bodies are small, removed them since other logs should give us enough info :+1: